File: src/me/bytebeats/algo/trees/BinaryTreeCodec.java
package me.bytebeats.algo.trees;

import me.bytebeats.algs.ds.TreeNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @Author bytebeats
 * @Email <happychinapc@gmail.com>
 * @Github https://github.com/bytebeats
 * @Created on 2021/6/30 16:50
 * @Version 1.0
 * @Description TO-DO
 */

public class BinaryTreeCodec {//剑指 Offer 37

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        if (root == null) {
            return "null:";
        }
        return root.val + ":" + serialize(root.left) + serialize(root.right);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] elements = data.split(":");
        List<String> list = new ArrayList();
        list.addAll(Arrays.asList(elements));
        return buildTree(list);
    }

    private TreeNode buildTree(List<String> elements) {
        if (elements == null || elements.isEmpty()) {
            return null;
        }
        String val = elements.remove(0);
        if ("null".equals(val)) {
            return null;
        }
        TreeNode root = new TreeNode(Integer.parseInt(val));
        root.left = buildTree(elements);
        root.right = buildTree(elements);
        return root;
    }
}

File: src/me/bytebeats/algorithms/kt/design/BinaryTreeCodec.kt
package me.bytebeats.algorithms.kt.design

import me.bytebeats.algorithms.meta.TreeNode

class BinaryTreeCodec { //297
    // Encodes a URL to a shortened URL.
    fun serialize(root: TreeNode?): String {
        if (root == null){
            return "null,"
        }
        val left = serialize(root.left)
        val right = serialize(root.right)
        return "${root.`val`},$left$right"
    }

    // Decodes your encoded data to tree.
    fun deserialize(data: String): TreeNode? {
        val element = data.split(",")
        return buildTree(element.toMutableList())
    }

    private fun buildTree(mutableList: MutableList<String>): TreeNode? {
        if (mutableList.isEmpty()) {
            return null
        }
        val e = mutableList.removeAt(0)
        if (e == "null") {
            return null
        }
        val root = TreeNode(e.toInt())
        root.left = buildTree(mutableList)
        root.right = buildTree(mutableList)
        return root
    }
}